File: exercism/ruby/binary-search/binary_search.rb
class BinarySearch
  attr_reader :list

  def initialize(list)
    raise ArgumentError unless list.each_cons(2).all?{|i, j| i <= j }
    @list = list
  end

  def middle
    @list.size / 2
  end

  def search_for(value, low = 0, high = @list.size - 1)
    raise RuntimeError if low > high

    pivot = (high + low) / 2

    case value <=> @list[pivot]
    when -1 then search_for(value, low, pivot - 1)
    when  0 then pivot
    when  1 then search_for(value, pivot + 1, high)
    end
  end
end

File: exercism/kotlin/binary-search/src/main/kotlin/BinarySearch.kt
object BinarySearch {
  fun search(xs: List<Int>, x: Int): Int {

    // This makes the search unusable, but is required by the specification T.T
    require((1 until xs.size).all { xs[it - 1] <= xs[it] }) { "Given unsorted array." }

    var low = 0
    var high = xs.size - 1
    while (low <= high) {
      var mid = (high + low) / 2
      when {
        xs[mid] > x -> high = mid - 1
        xs[mid] < x -> low = mid + 1
        xs[mid] == x -> return mid
      }
    }
    return -1
  }
}
